src/GBACartMotionPak.cpp

```diff
@@ -0,0 +1,196 @@
+    the terms of the GNU General Public License as published by the Free
+    Software Foundation, either version 3 of the License, or (at your option)
+    any later version.
+
+    melonDS is distributed in the hope that it will be useful, but WITHOUT ANY
+    WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+    FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License along
+    with melonDS. If not, see http://www.gnu.org/licenses/.
+*/
+
+#include <assert.h>
+#include "NDS.h"
+#include "GBACart.h"
+#include "Platform.h"
+#include <algorithm>
+#include "math.h"
+
+namespace melonDS
+{
+using Platform::Log;
+using Platform::LogLevel;
+
+namespace GBACart
+{
+
+CartMotionPakHomebrew::CartMotionPakHomebrew(void* userdata) : 
+    CartCommon(MotionPakHomebrew),
+    UserData(userdata)
+{
+}
+
+CartMotionPakHomebrew::~CartMotionPakHomebrew() = default;
+
+void CartMotionPakHomebrew::Reset()
+{
+    ShiftVal = 0;
+}
+
+void CartMotionPakHomebrew::DoSavestate(Savestate* file)
+{
+    CartCommon::DoSavestate(file);
+    file->Var16(&ShiftVal);
+}
+
+u16 CartMotionPakHomebrew::ROMRead(u32 addr) const
+{
+    // CHECKME: Does this apply to the homebrew cart as well?
+    return 0xFCFF;
+}
+
+static int AccelerationToMotionPak(float accel)
+{
+    const float GRAVITY_M_S2 = 9.80665f;
+
+    return std::clamp(
+        (int) ((accel / (5 * GRAVITY_M_S2) + 0.5) * 4096),
+        0, 4095
+    );
+}
+
+static int AccelerationToMotionPakRetail(float accel)
+{
+    const float GRAVITY_M_S2 = 9.80665f;
+
+    return std::clamp(
+        (int) ((accel / (5 * GRAVITY_M_S2) + 0.5) * 256),
+        0, 254
+    );
+}
+
+static int RotationToMotionPak(float rot)
+{
+    const float DEGREES_PER_RAD = 180 / M_PI;
+    const float COUNTS_PER_DEG_PER_SEC = 0.825;
+    const int CENTER = 1680;
+
+    return std::clamp(
+        (int) ((rot * DEGREES_PER_RAD * COUNTS_PER_DEG_PER_SEC) + CENTER + 0.5),
+        0, 4095
+    );
+}
+
+u8 CartMotionPakHomebrew::SRAMRead(u32 addr)
+{
+    // CHECKME: SRAM address mask
+    addr &= 0xFFFF;
+
+    switch (addr)
+    {
+    case 0:
+        // Read next byte
+        break;
+    case 2:
+        // Read X acceleration
+        ShiftVal = AccelerationToMotionPak(Platform::Addon_MotionQuery(Platform::MotionAccelerationX, UserData)) << 4;
+        // CHECKME: First byte returned when reading acceleration/rotation
+        return 0;
+    case 4:
+        // Read Y acceleration
+        ShiftVal = AccelerationToMotionPak(Platform::Addon_MotionQuery(Platform::MotionAccelerationY, UserData)) << 4;
+        return 0;
+    case 6:
+        // Read Z acceleration
+        ShiftVal = AccelerationToMotionPak(Platform::Addon_MotionQuery(Platform::MotionAccelerationZ, UserData)) << 4;
+        return 0;
+    case 8:
+        // Read Z rotation
+        // CHECKME: This is a guess, compare with real hardware
```

Comment:
You may have a hard time here as I'm not entirely convinced the homebrew Slot2 motion pack was ever actually sold/made available. The only product I can find with a motion pack is supposedly the Neoflash "R6Gold".

All of my knowledge of the homebrew cart's spec comes from libnds.

Reply:
> You may have a hard time here as I'm not entirely convinced the homebrew Slot2 motion pack was ever actually sold/made available.

Absolutely was. https://web.archive.org/web/20080509173200/http://www.ndsmotion.com/ has a list of stores and photos of the packaging.

There were six products in total which featured Kionix's technology, five of which were for the DS family of consoles; I maintain a list here: https://wiki.asie.pl/doku.php?id=notes:flashcart:ds_motion_card

I tested with an MK6-Motion myself, which happens to feature both the accelerometer and gyroscope chip; since the chips themselves and thus values output should be identical between the Card and a Pak, I deemed the emulation to be accurate enough for running homebrew (which is the important thing for me from a preservation perspective), but I can't verify fine details like the Pak's open bus behaviour or any additional SPI/ROM commands the Card might have.

> All of my knowledge of the homebrew cart's spec comes from libnds.

There's some additional bits of information in a README bundled with the original DS Motion Card/Pak driver source code on said website.

Reply:
(Either way, I don't expect this to be resolved as part of this PR, and the detection mechanism used by all homebrew for the Pak doesn't rely on open bus behaviour.)